packages/imitation_game/README.md

```diff
@@ -0,0 +1,89 @@
+# Imitation Game
+
+## Description
+
+`imitation_game` is a platform for performing automated tests to compare Flutter
+to alternative platforms.  For example, how much memory does the same app
+written in Flutter and UIKit take?
+
+## Running all the tests
+
+You need a mobile device plugged into your computer and setup for development.
+The mobile device and the computer need to be on the same network, one that
+allows communication between computers since that's how the mobile phone will
+report its results to the computer.
+
+```sh
+dart imitation_game.dart
+```
+
+## Dependencies
+
+In order to run the tests you will need the union of all the platforms being
+tested.  As new tests are added please add to this list:
+
+### iOS
+
+- Flutter
+- Xcode
+- [ios_deploy](https://github.com/ios-control/ios-deploy) - used to launch apps
+  on the attached iOS device.
+
+## File Layout
+
+```
+./
+├─ imitation_game.dart
+└─ tests/
+   ├─ smiley/
+   │  ├─ README.md
+   │  ├─ flutter/
+   │  │  └─ run_ios.sh
+   │  └─ uikit/
+   │     └─ run_ios.sh
+   └─ memory/
+      ├─ README.md
+      ├─ flutter/
+      │  └─ run_ios.sh
+      └─ uikit/
+         └─ run_ios.sh
+```
+
+Here there are 2 different tests with 2 different platform implementations.  The
+tests are named `smiley` and `memory`, they are both implemented on the
+platforms `flutter` and `uikit`.
+
+### Adding a test
+
+Tests should comprise of implementations on one or more platform.  The directory
+for the test should be added to `./tests`.  Inside that directory there should
+be a directory of implementations and a `README.md` file that explains the test.
+
+### Adding an implementation to a test
+
+An implementation has to follow these rules:
+
+- It needs to perform the same operations as the other implementations and
+  follow the description in the test's `README.md`.
+- It needs to contain a `run_ios.sh` script that will build and launch the test
+  on the connected device.
+- It should contain a file named `ip.txt` which will be overwritten by
+  `imitation_game.dart` with the ip address and port that should be used to
+  report results to.
+- It needs to report its results to the ip and port in the `ip.txt` via an HTTP
+  POST of JSON data.
```

Comment:
I feel that the measurement code/tool should not live inside the implementation directory. There should only be a single measurement code/tool that's under the test (e.g., `smiley`) directory, and all implementations must share it to measure the performance. Otherwise, it's hard to prove that different implementations are measuring the same thing.

This might be critical when a UI framework that no one is familiar enters the arena. Say if someone invented a new framework called Glutter, we probably don't have the expertise and time to vet its measuring code and tools.

Reply:
We just have to rely on humans to read the code.  Someone reading the code can say if it is measuring the same thing.  Certain things are only measurable on device, like time since the process was launched.  From the host computer we can just measure the time between when we asked it to launch which isn't as accurate.

If we find there is a way to share measurement code that can be factored out.  Nothing about the current design disallows it.